std/lazyt.glu

```diff
@@ -0,0 +1,34 @@
+//@NO-IMPLICIT-PRELUDE
+
+let { (<<) } = import! std.function
+let { lazy, force } = import! std.lazy
+let { Applicative, Functor, Monad } = import! std.prelude
+
+type LazyT m a = Lazy (m a)
+
+let functor ?fr : [Functor m] -> Functor (LazyT m) =
```

Comment:
I get odd errors when attempting to omit `?fr`, `?ap`, and such in `lazyt.glu` and `statet.glu`. Here's the error removing `?fr` from the above line causes:

```
error: Expected the following types to be equal
Expected: Lazy a0
Found: a0
1 errors were found during unification:
Variable `11` occurs in `Lazy a0`.
- <std.lazyt>:12:39
12 |     let map f ma = lazy (\_ -> map f (force ma))

error: Expected the following types to be equal
Expected: a1
Found: Lazy a1
1 errors were found during unification:
Variable `5` occurs in `Lazy a1`.
- <std.lazyt>:12:20
12 |     let map f ma = lazy (\_ -> map f (force ma))
```

All the other errors seem similar in nature.

Reply:
Ah, the issue is that the `map` you define is the one you end up calling inside it as well

```f#
let map f ma = lazy (\_ -> map f (force ma))
 // ^                      ^
 // same
```

But inside the `map` you want to call the one imported from `std.functor`. Rename either of them and it ought to work.

Reply:
Ah good, thanks. That was a simple fix. But should names be available in their own definition bodies outside of recursive contexts?

Reply:
`rec let` was added only a few months ago as a way to define recursive values. Before that only functions could be recursive and they were always recursive by default. (since there were no way to  to mark them as recursive/non-recursive).

In the PR that added `rec let` I kept the recursive by default behaviour to avoid a breaking changes while `rec let` was still experimental. Making them only recursive when explicitly using `rec let` may be a good idea though.

Reply:
https://github.com/gluon-lang/gluon/issues/692